i3lock.c

```diff
@@ -645,6 +645,120 @@ void handle_screen_resize(void) {
     redraw_screen();
 }
 
+static int read_raw_image_native(uint32_t *dest, FILE *src, int width, int height, int stride) {
+    int count = 0;
+    for (int y = 0; y < height; y++) {
+        size_t n = fread(&dest[y * stride], 1, width * 4, src);
+        count += n;
+        if (n < (size_t)(width * 4))
+            break;
+    }
+
+    return count;
+}
+
+static int read_raw_image_rgb(uint32_t *dest, FILE *src, int width, int height, int stride) {
+    unsigned char *buf = malloc(width * 3);
+    if (buf == NULL)
+        return -1;
+
+    int count = 0;
+    for (int y = 0; y < height; y++) {
+        size_t n = fread(buf, 1, width * 3, src);
+        count += n;
+        if (n < (size_t)(width * 3))
+            break;
+
+        for (int x = 0; x < width; ++x) {
+            int idx = x * 3;
+            dest[y * stride + x] = 0 |
+                                   (buf[idx + 0] << 16) |
+                                   (buf[idx + 1] << 8) |
+                                   (buf[idx + 2]);
+        }
+    }
+
+    free(buf);
+    return count;
+}
+
+static cairo_surface_t *read_raw_image(const char *image_path, const char *image_raw_format) {
+    cairo_surface_t *img;
+
+    /* Parse format as <width>x<height>[:<format>] */
+    int w = -1, h = -1;
+    char format[] = "native";
+    char *fmtstring = NULL;
+    if (asprintf(&fmtstring, "%%ix%%i:%%%lds", sizeof(format) - 1) == -1) {
+        errx(EXIT_FAILURE, "asprintf() failed.");
+    }
+
+    if (sscanf(image_raw_format, fmtstring, &w, &h, format) == EOF || w == -1 || h == -1) {
+        fprintf(stderr, "Invalid image format \"%s\"\n", image_raw_format);
+        free(fmtstring);
+        return NULL;
+    }
+
+    free(fmtstring);
+
+    /* Create image surface */
+    img = cairo_image_surface_create(CAIRO_FORMAT_RGB24, w, h);
+    if (cairo_surface_status(img) != CAIRO_STATUS_SUCCESS) {
+        fprintf(stderr, "Could not create surface: %s\n",
+                cairo_status_to_string(cairo_surface_status(img)));
+        return NULL;
+    }
+    cairo_surface_flush(img);
+
+    /* Use uint32_t* because cairo uses native endianness */
+    uint32_t *data = (uint32_t *)cairo_image_surface_get_data(img);
+    const int stride = cairo_image_surface_get_stride(img) / 4;
```

Comment:
Uhm. It surprises me that you and cairo use different definitions of "stride". I know that this has to be divided by four due to C, but... still, this surprises me here.

Reply:
I renamed it to pixstride to make it less confusing. That seems to be what cairo internally uses when talking about a stride in terms of pixels rather than bytes.